### ISSUE TITLE
ssh_gateway ignored

### ISSUE THREAD

**Issue by kplimack** [chef/chef-provisioning#481]
I'm trying to tunnel ssh connections through a bastion host, but `ssh_gateway` seems to be ignored no matter where I try specifying it.

```
[5] pry(#<Chef::Provisioning::AWSDriver::Driver>)> machine_options
=> #<Cheffish::MergedConfig:0x007fdc0ceb7f40
 @configs=
  [{:convergence_options=>
     {:chef_server=>{:chef_server_url=>"chefzero://localhost:8889", :options=>{:client_name=>"Jakes-MacBook-Pro.local", :signing_key_filename=>nil, :api_version=>"0"}},
      :allow_overwrite_keys=>nil,
      :source_key=>nil,
      :source_key_path=>nil,
      :source_key_pass_phrase=>nil,
      :private_key_options=>nil,
      :ohai_hints=>nil,
      :public_key_path=>nil,
      :public_key_format=>nil,
      :admin=>nil,
      :validator=>nil,
      :chef_config=>nil}},
   {:bootstrap_options=>
     {:subnet=>"us-east-1d-public",
      :image_id=>"ami-68115b02",
      :key_name=>"kplimack",
      :key_path=>"/Users/kplimack/.ssh/kplimack-aws.pem",
      :security_group_ids=>
       [<aws_security_group[default-sg-vpc] @name: "default-sg-vpc" @noop: nil @before: nil @params: {} @provider: nil @allowed_actions: [:nothing, :create, :destroy, :purge] @action: [:create] @updated: false @updated_by_last_action: false @supports: {} @ignore_failure: false @retries: 0 @retry_delay: 2 @source_line: "/Users/kplimack/repos/myOrg-aws-pipeline/.chef/local-mode-cache/cache/cookbooks/myOrg-aws-pipeline/recipes/vpc_us-east-1.rb:46:in `block (2 levels) in from_file'" @guard_interpreter: nil @default_guard_interpreter: :default @elapsed_time: 1.897933 @sensitive: false @driver: #<Chef::Provisioning::AWSDriver::Driver:0x007fdc135e89a8 @driver_url="aws::us-east-1", @config={:log_location=>#<IO:<STDOUT>>, :cookbook_path=>"/Users/kplimack/repos/myOrg-aws-pipeline/.chef/local-mode-cache/cache/cookbooks", :force_logger=>false, :force_formatter=>false, :color=>true, :named_run_list=>nil, :log_level=>:debug, :local_mode=>true, :override_runlist=>[#<Chef::RunList::RunListItem:0x007fdc13028798 @version=nil, @type=:recipe, @name="myOrg-aws-pipeline::vpc_us-east-1">], :config_file=>"/Users/kplimack/repos/myOrg-aws-pipeline/.chef/config.rb", :specific_recipes=>[], :chef_server_url=>"chefzero://localhost:8889", :repo_mode=>"everything", :formatters=>[], :event_handlers=>[], :node_name=>"Jakes-MacBook-Pro.local", :start_handlers=>[], :live_stream=>true, :chef_provisioning=>{:machine_max_wait_time=>500}, :user_valid_regex=>[/^[^-+~:,\t\r\n\f\0]+[^:,\t\r\n\f\0]*$/], :group_valid_regex=>[/^[^-+~:,\t\r\n\f\0]+[^:,\t\r\n\f\0]*$/]}, @aws_credentials=#<Chef::Provisioning::AWSDriver::Credentials:0x007fdc135e8890 @credentials={"default"=>{:aws_access_key_id=>"AWS_ACCESS_KEY", :aws_secret_access_key=>"AWS_SECRET_KEY"}}>, @aws_config=<AWS::Core::Configuration>, @aws_config_2={:credentials=>#<Aws::Credentials access_key_id="AWS_ACCESS_KEY">, :region=>"us-east-1", :http_proxy=>nil, :logger=>#<MonoLogger:0x007fdc13023b80 @progname=nil, @level=0, @default_formatter=#<Logger::Formatter:0x007fdc13023c98 @datetime_format=nil>, @formatter=#<Mixlib::Log::Formatter:0x007fdc1301bed0 @datetime_format=nil>, @logdev=#<MonoLogger::LocklessLogDevice:0x007fdc13023dd8 @shift_size=nil, @shift_age=nil, @filename=nil, @dev=#<IO:<STDOUT>>>>, :retry_limit=>5}, @ec2=<AWS::EC2>, @ec2_client=#<Aws::EC2::Client>, @ec2_resource=#<Aws::EC2::Resource>, @route53_client=#<Aws::Route53::Client>> @chef_server: {:chef_server_url=>"chefzero://localhost:8889", :options=>{:client_name=>"Jakes-MacBook-Pro.local", :signing_key_filename=>nil, :api_version=>"0"}} @declared_type: :aws_security_group @cookbook_name: "myOrg-aws-pipeline" @recipe_name: "vpc_us-east-1" @vpc: #<Chef::DelayedEvaluator:0x007fdc13610d18@/Users/kplimack/repos/myOrg-aws-pipeline/.chef/local-mode-cache/cache/cookbooks/myOrg-aws-pipeline/recipes/vpc_us-east-1.rb:47> @inbound_rules: [{:port=>22..22, :protocol=>:tcp, :sources=>["0.0.0.0/0"]}, {:port=>-1..-1, :protocol=>:icmp, :sources=>["0.0.0.0/0"]}] @outbound_rules: [{:port=>-1..-1, :protocol=>:any, :destinations=>["0.0.0.0/0"]}] @description: "Default VPC Security Group" @managed_entry_store: #<Chef::Provisioning::ChefManagedEntryStore:0x007fdc13b44ac8 @chef_server={:chef_server_url=>"chefzero://localhost:8889", :options=>{:client_name=>"Jakes-MacBook-Pro.local", :signing_key_filename=>nil, :api_version=>"0"}}> @aws_object_id: "sg-c941afb0">],
      :associate_public_ip_address=>false,
      :ssh_gateway=>"ec2-user@BASTION_IP"},
    :machine_options=>{:ssh_username=>"ec2-user"},
    :transport_options=>{:ssh_options=>{:proxy=>"ec2-user@BASTION_IP"}, :options=>{:ssh_gateway=>"ec2-user@BASTION_IP"}, :ssh_gateway=>"ec2-user@BASTION_IP"},
    :ssh_gateway=>"admin@BASTION_IP"}],
 @merge_arrays={}>
[6] pry(#<Chef::Provisioning::AWSDriver::Driver>)> ^D
[2015-12-21T15:44:34-07:00] INFO: Executing sudo pwd on ec2-user@192.168.0.6
[2015-12-21T15:44:34-07:00] DEBUG: Opening SSH connection to ec2-user@192.168.0.6 with options {:timeout=>10, :auth_methods=>["publickey"], :keys_only=>true, :host_key_alias=>"i-d0819160.AWS"}
[2015-12-21T15:44:44-07:00] DEBUG: Timed out connecting to SSH: execution expired
[2015-12-21T15:44:44-07:00] DEBUG: ec2-user@192.168.0.6 unavailable: network connection failed or broke: #<Chef::Provisioning::Transport::SSH::InitialConnectTimeout: execution expired>

    - waiting for nat-us-east-1d (i-d0819160 on aws::us-east-1) to be connectable (transport up and running) ...
    - been waiting 0/500 -- sleeping 10 seconds for nat-us-east-1d (i-d0819160 on aws::us-east-1) to become connectable ...[2015-12-21T15:44:44-07:00] INFO: Executing sudo pwd on ec2-user@192.168.0.6
[2015-12-21T15:44:44-07:00] DEBUG: Opening SSH connection to ec2-user@192.168.0.6 with options {:timeout=>10, :auth_methods=>["publickey"], :keys_only=>true, :host_key_alias=>"i-d0819160.AWS"}
^C[2015-12-21T15:44:47-07:00] FATAL: SIGINT received, stopping
[2015-12-21T15:44:47-07:00] FATAL: SIGINT received, stopping
rake aborted!
```

```
      bastion = search(:node, "name:bastion-#{zone}").first
      bastion_ipaddress = bastion['ec2']['public_ipv4']

      machine "nat-#{zone}" do
        machine_options bootstrap_options: {
                          subnet: "#{zone}-public",
                          image_id: ec2_amazon_nat_ami,
                          key_name: u,
                          key_path: private_key,
                          security_group_ids: [security_group_default],
                          associate_public_ip_address: false,
                          ssh_gateway: "#{ec2_amazon_nat_username}@#{bastion_ipaddress}"
                        },
                        machine_options: {
                          ssh_username: ec2_amazon_nat_username,
                          ssh_gateway: "#{ec2_amazon_nat_username}@#{bastion_ipaddress}"
                        },
                        transport_options: {
                          ssh_options: {
                            proxy: "#{ec2_amazon_nat_username}@#{bastion_ipaddress}",
                            ssh_gateway: "#{ec2_amazon_nat_username}@#{bastion_ipaddress}"
                          },
                          options: {
                            ssh_gateway: "#{ec2_amazon_nat_username}@#{bastion_ipaddress}"
                          },
                          ssh_gateway: "#{ec2_amazon_nat_username}@#{bastion_ipaddress}"
                        },
                        ssh_gateway: "#{ec2_debian_jessie_username}@#{bastion_ipaddress}" # this should use ENV['AWS_USER'] once the
bastion has user's keys bootstrapped
      end
```

**Comment by kplimack** [comment]
@tyler-ball do you know what I'm missing here?

**Comment by kplimack** [comment]
@jkeiser ?

**Comment by ngourley** [comment]
I thought I had this problem too, but when the use of the gateway fails, it displays the host information for the provisioned host instead of the gateway host. I think this makes debugging a little difficult, I ended up putting debug statements in the gem source and doing trail runs on new hosts. For it to work for me, the account in ssh_gateway had to match the ssh_username. I assume this is because both hosts were provisioned with and use the same keys. 
My config:
```
with_machine_options({
  bootstrap_options: {
    flavor_ref: bootstrap_options['flavor_ref']['small'],
    image_ref: bootstrap_options['image_ref']['7.1'],
    key_name: bootstrap_options['key_name'],
    # floating_ip_pool: bootstrap_options['floating_ip_pool'],
    nics: [{ :net_id => bootstrap_options['nics']['net_id']}]
  },
  ssh_username: 'cloud-user',
  ssh_gateway: 'cloud-user@10.190.188.120',
})
```

Are the accounts you are using the same on both hosts?
Documentation:
```
  ssh_username: "ubuntu", # Username to use for ssh and WinRM
  ssh_gateway: "user@gateway", " SSH gateway configuration
  ssh_options: { # a list of options to Net::SSH.start
    :auth_methods => [ 'publickey' ], # DEFAULT
    :keys_only => true, # DEFAULT
    :host_key_alias => "#{instance.id}.AWS" # DEFAULT
  }
```

Hope this helps!

**Comment by kplimack** [comment]
thanks nathan.
in this case, i'm setting up 2 boxes.  first a bastion, and second a NAT
instance between two subnets.
once the NAT is instanciated, i want to connect via the bastion.  the
bastion uses my personal credentials where the NAT uses ec2-user,
so my ssh-gatweway should me me@bastion, ssh_username should be me... i
think.?

**Comment by ngourley** [comment]
```ruby
# Assumptions from your comment
# bastion uses account 'me' aka your account, which is the ssh gateway
# the NAT host uses 'ec2-user', which is the host being provisioned
bastion = search(:node, "name:bastion-#{zone}").first
bastion_ipaddress = bastion['ec2']['public_ipv4']

bootstrap_options: {
  ssh_username: "ec2-user",
  ssh_gateway: "me@#{bastion_ipaddress}"
}
```
I think we are on the same page here. But, I think the problem is that after initial provisioning, the bastion host cannot SSH to the provisioned host. Can you login to the provisioned host simply by doing ssh ec2-user@<host ip> from the bastion host? I think probably it won't.

It looks like the "ssh_options" and setting the location of the correct identity file might be the next step. Can you give that a shot?
https://github.com/chef/chef-provisioning/blob/master/README.md#machine-options
https://github.com/net-ssh/net-ssh/blob/master/lib/net/ssh.rb#L136

**Comment by kplimack** [comment]
@ngourley I tailed the auth log and tcpdump'd on the bastion, but did not see the bastion being used as a gateway to reach the NAT instance.  

```
[5] pry(#<Chef::Provisioning::AWSDriver::Driver>)> machine_options[:bootstrap_options].keys
=> [:subnet, :image_id, :key_name, :key_path, :security_group_ids, :associate_public_ip_address, :iam_instance_profile, :ssh_gateway]
[6] pry(#<Chef::Provisioning::AWSDriver::Driver>)> machine_options[:bootstrap_options][:ssh_gateway]
=> "admin@BASTION_IPADDRESS"
[7] pry(#<Chef::Provisioning::AWSDriver::Driver>)> ^D
[2016-01-11T15:26:00-07:00] INFO: Executing sudo pwd on ami-68115b02@192.168.0.6
[2016-01-11T15:26:00-07:00] DEBUG: Opening SSH connection to ami-68115b02@192.168.0.6 with options {:timeout=>10, :auth_methods=>["publickey"], :keys_only=>true, :host_key_alias=>"i-d0819160.AWS"}
[2016-01-11T15:26:10-07:00] DEBUG: Timed out connecting to SSH: execution expired
[2016-01-11T15:26:10-07:00] DEBUG: ami-68115b02@192.168.0.6 unavailable: network connection failed or broke: #<Chef::Provisioning::Transport::SSH::InitialConnectTimeout: execution expired>

    - waiting for nat-us-east-1d (i-d0819160 on aws::us-east-1) to be connectable (transport up and running) ...
    - been waiting 0/500 -- sleeping 10 seconds for nat-us-east-1d (i-d0819160 on aws::us-east-1) to become connectable ...[2016-01-11T15:26:10-07:00] INFO: Executing sudo pwd on ami-68115b02@192.168.0.6
[2016-01-11T15:26:10-07:00] DEBUG: Opening SSH connection to ami-68115b02@192.168.0.6 with options {:timeout=>10, :auth_methods=>["publickey"], :keys_only=>true, :host_key_alias=>"i-d0819160.AWS"}
[2016-01-11T15:26:20-07:00] DEBUG: Timed out connecting to SSH: execution expired
[2016-01-11T15:26:20-07:00] DEBUG: ami-68115b02@192.168.0.6 unavailable: network connection failed or broke: #<Chef::Provisioning::Transport::SSH::InitialConnectTimeout: execution expired>
^C[2016-01-11T15:26:21-07:00] FATAL: SIGINT received, stopping
[2016-01-11T15:26:21-07:00] FATAL: SIGINT received, stopping
rake aborted!
Interrupt:
```

**Comment by ngourley** [comment]
Can you post the code you added to debug?

**Comment by kplimack** [comment]
@ngourley 
I added a `pry` to `/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-provisioning-aws-1.7.0/lib/chef/provisioning/aws_driver/driver.rb`

```
    def create_ssh_transport(machine_spec, machine_options, instance)
      ssh_options = ssh_options_for(machine_spec, machine_options, instance)
      username = machine_spec.reference['ssh_username'] || machine_options[:ssh_username] || default_ssh_username
      if defined?(machine_options[:machine_options]) && defined?(machine_options[:machine_options][:ssh_username]) && username ==
default_ssh_username
        username = machine_options[:machine_options][:ssh_username]
      end

      if machine_options.has_key?(:ssh_username) && machine_options[:ssh_username] != machine_spec.reference['ssh_username']
        Chef::Log.warn("Server #{machine_spec.name} was created with SSH username #{machine_spec.reference['ssh_username']} and
machine_options specifies username #{machine_options[:ssh_username]}.  Using #{machine_spec.reference['ssh_username']}.  Please edit
the node and change the chef_provisioning.reference.ssh_username attribute if you want to change it.")
      end
      options = {}
      if machine_spec.reference[:sudo] || (!machine_spec.reference.has_key?(:sudo) && username != 'root')
        options[:prefix] = 'sudo '
      end

      remote_host = determine_remote_host(machine_spec, instance)

      #Enable pty by default
      options[:ssh_pty_enable] = true
      options[:ssh_gateway] = machine_spec.reference['ssh_gateway'] if machine_spec.reference.has_key?('ssh_gateway')
      if machine_spec.name.include?('nat')
        require 'pry'; binding.pry
      end

      Chef::Provisioning::Transport::SSH.new(remote_host, username, ssh_options, options, config)
    end
```

**Comment by kplimack** [comment]
@ngourley do you know where machine_spec.reference is supposed to be getting set? seems like `machine_spec.reference['ssh_gateway']` is not set despite the many places im attempting to do so.

```
    1025:   #Enable pty by default
    1026:   options[:ssh_pty_enable] = true
    1027:   options[:ssh_gateway] = machine_spec.reference['ssh_gateway'] if machine_spec.reference.has_key?('ssh_gateway')
    1028:   if machine_spec.name.include?('nat')
 => 1029:     require 'pry'; binding.pry
    1030:   end
    1031:
    1032:   Chef::Provisioning::Transport::SSH.new(remote_host, username, ssh_options, options, config)
    1033: end

[1] pry(#<Chef::Provisioning::AWSDriver::Driver>)> options
=> {:prefix=>"sudo ", :ssh_pty_enable=>true}
[2] pry(#<Chef::Provisioning::AWSDriver::Driver>)> machine_spec.reference
=> {"driver_version"=>"1.7.0",
 "allocated_at"=>"2015-12-17 05:24:15 UTC",
 "host_node"=>"chefzero://localhost:8889/nodes/",
 "image_id"=>"ami-68115b02",
 "instance_id"=>"i-d0819160",
 "key_name"=>"kplimack"}
```

**Comment by kplimack** [comment]
@ngourley i seem to have gotten it to work by adding that line on 1128

```
    1025:   #Enable pty by default
    1026:   options[:ssh_pty_enable] = true
    1027:   options[:ssh_gateway] = machine_spec.reference['ssh_gateway'] if machine_spec.reference.has_key?('ssh_gateway')
    1028:   options[:ssh_gateway] = machine_options[:machine_options][:ssh_gateway] if machine_options[:machine_options].has_key?(:ssh_gateway) and options[:ssh_gateway].nil?
    1029:
    1030:   if machine_spec.name.include?('nat')
    1031:
 => 1032:     require 'pry'; binding.pry
    1033:   end
    1034:
    1035:   Chef::Provisioning::Transport::SSH.new(remote_host, username, ssh_options, options, config)
    1036: end

[1] pry(#<Chef::Provisioning::AWSDriver::Driver>)> options
=> {:prefix=>"sudo ", :ssh_pty_enable=>true, :ssh_gateway=>"admin@xxx.yyy.zzz.xxx"}
[2] pry(#<Chef::Provisioning::AWSDriver::Driver>)> ^D
[2016-01-12T16:09:01-08:00] INFO: Executing sudo pwd on ami-68115b02@192.168.0.6
[2016-01-12T16:09:01-08:00] DEBUG: Opening SSH connection to ami-68115b02@192.168.0.6 with options {:timeout=>10, :auth_methods=>["publickey"], :keys_only=>true, :host_key_alias=>"i-d0819160.AWS"}
[2016-01-12T16:09:01-08:00] DEBUG: Opening SSH gateway to admin@xxx.yyy.xxx.zzz with options {:timeout=>10, :auth_methods=>["publickey"], :keys_only=>true, :host_key_alias=>"i-d0819160.AWS", :port=>22}
[2016-01-12T16:09:02-08:00] DEBUG: ami-68115b02@192.168.0.6 unavailable: SSH authentication error: #<Net::SSH::AuthenticationFailed: Authentication failed for user ami-68115b02@127.0.0.1>
```

**Comment by ngourley** [comment]
I have the same question that you do, what is machine_spec and how is it
different than machine_options? Also what is options from line 1028? Can
you log out these values? I'm glad you found a way to make it work perhaps
it is a setting missing after all.

On Tue, Jan 12, 2016 at 6:10 PM, Jake Plimack <notifications@github.com>
wrote:

> @ngourley <https://github.com/ngourley> i seem to have gotten it to work
> by adding that line on 1128
>
>     1025:   #Enable pty by default
>     1026:   options[:ssh_pty_enable] = true
>     1027:   options[:ssh_gateway] = machine_spec.reference['ssh_gateway'] if machine_spec.reference.has_key?('ssh_gateway')
>     1028:   options[:ssh_gateway] = machine_options[:machine_options][:ssh_gateway] if machine_options[:machine_options].has_key?(:ssh_gateway) and options[:ssh_gateway].nil?
>     1029:
>     1030:   if machine_spec.name.include?('nat')
>     1031:
>  => 1032:     require 'pry'; binding.pry
>     1033:   end
>     1034:
>     1035:   Chef::Provisioning::Transport::SSH.new(remote_host, username, ssh_options, options, config)
>     1036: end
>
> [1] pry(#<Chef::Provisioning::AWSDriver::Driver>)> options
> => {:prefix=>"sudo ", :ssh_pty_enable=>true, :ssh_gateway=>"admin@xxx.yyy.zzz.xxx"}
> [2] pry(#<Chef::Provisioning::AWSDriver::Driver>)> ^D
> [2016-01-12T16:09:01-08:00] INFO: Executing sudo pwd on ami-68115b02@192.168.0.6
> [2016-01-12T16:09:01-08:00] DEBUG: Opening SSH connection to ami-68115b02@192.168.0.6 with options {:timeout=>10, :auth_methods=>["publickey"], :keys_only=>true, :host_key_alias=>"i-d0819160.AWS"}
> [2016-01-12T16:09:01-08:00] DEBUG: Opening SSH gateway to admin@xxx.yyy.xxx.zzz with options {:timeout=>10, :auth_methods=>["publickey"], :keys_only=>true, :host_key_alias=>"i-d0819160.AWS", :port=>22}
> [2016-01-12T16:09:02-08:00] DEBUG: ami-68115b02@192.168.0.6 unavailable: SSH authentication error: #<Net::SSH::AuthenticationFailed: Authentication failed for user ami-68115b02@127.0.0.1>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/chef/chef-provisioning/issues/481#issuecomment-171106540>
> .
>



-- 
Nathan Gourley
ngourley@gmail.com

**Comment by kplimack** [comment]
@ngourley you can see the values in the output above.
```
[1] pry(#<Chef::Provisioning::AWSDriver::Driver>)> options
=> {:prefix=>"sudo ", :ssh_pty_enable=>true, :ssh_gateway=>"admin@xxx.yyy.zzz.xxx"}
```

**Comment by jgoulah** [comment]
@kplimack fwiw you're setting the machine options incorrectly, so I'm not sure its right to patch the code to alleviate this (and to be fair, some chef provisioning documentation is showing this incorrect way which is even more confusing)

it should look like this, and this works to pass the ssh_gateway:

```
     machine_options(
       :bootstrap_options => {
          :key_name => 'XXX,
          :key_path => "#{ENV['HOME']}/.ssh/",
          # other boostrap options here
        },
        # now these are your machine options
        :ssh_gateway => "ubuntu@bastion",
        :transport_address_location => :private_ip,
      )
```

**Comment by kplimack** [comment]
thanks @jgoulah, i'll give that shot.

**Comment by tyler-ball** [comment]
@kplimack Yeah, looking at your initial report, I see that you were setting `ssh_username` and `ssh_gateway` at the wrong level. They should be at the top level of the machine options, at the same lever `bootstrap_options` is specified.

An explanation of the `machine_spec.reference`: The very first time the machine resource is converged it calls the `create_isntances_and_references` method. In there [we copy](https://github.com/chef/chef-provisioning-aws/blob/63757067a4128e9212ab9189078d5476df3bb6db/lib/chef/provisioning/aws_driver/driver.rb#L1480) certain machine_options into the `machine_spec.reference`.

The purpose of `machine_spec.reference` is to persist information needed to connect to the instance in AWS. For example, we store the instance id and the key name in the `machine_spec.reference`. The reason we persist this is to support something like the following:

```ruby
machine 'my_awesome_machine' do
  machine_options({
    bootstrap_options: { ... },
    ssh_username: 'centos'
  })
  action :allocate
end

#... A bunch of other stuff, like setting up a bastion host

# The first time we just allocated the machine, now we are actually going to converge it
machine 'my_awesome_machine' do
  machine_options({ convergence_options: { chef_version: "12.11" } })
  action :converge
end
```

Even though the second invocation of the `machine` resource does not include the `ssh_username` we are still able to successfully connect and converge because we persisted `ssh_username` into the `machine_spec.reference`. That is why in [`create_ssh_transport`](https://github.com/chef/chef-provisioning-aws/blob/63757067a4128e9212ab9189078d5476df3bb6db/lib/chef/provisioning/aws_driver/driver.rb#L1153) we first check to see if the `ssh_username` is persisted into the `machine_spec.reference` before looking in the `machine_options`.

I'm of the opinion that we should first look in the `machine_options` then look in the `machine_spec.reference`, but that is a non-backwards compatible change that I don't have time to tackle today.

TL;DR: I'm going to put out a commit that looks for the `ssh_gateway` in the machine_options _and_ the `machine_spec.reference`, with preference given to the `machine_spec.reference`.
